### ISSUE TITLE
libnss_resolve.so crash: gethostbyaddr with link-local address of disconnected adapter

### ISSUE THREAD

**Issue by WGH-** [systemd/systemd#2785]
In order to reproduce the crash, you need to do the following:
- Enable resolve in `/etc/nsswitch.conf` (e.g. `hosts:       files resolve mdns`)
- Have a wired network adapter
- Disconnect the cable from the adapter
- Call gethostbyaddr with IPv6 link-local address of disconnected adapter

```
% ip address show dev enp0s25 | grep inet6
    inet6 fe80::xxxx:xxxx:xxxx:xxxx/64 scope link
% python3 -c "import socket; socket.gethostbyaddr('fe80::xxxx:xxxx:xxxx:xxxx%enp0s25')"
zsh: segmentation fault  python3 -c
```

In Python, implementation of `socket.getfqdn` calls `gethostbyaddr` with one of the local addresses, which might be the said address of disconnected adapter.
So as of me, `import socket; socket.getfqdn()` segfaults as well (that's how I initially discovered this).

I'm running systemd 228 on Gentoo Linux.

**Comment by keszybz** [comment]
I cannot reproduce this. Any chance you could capture a traceback or provide a core file?

**Comment by WGH-** [comment]
Okay, I was slightly wrong. libnss_resolve itself doesn't crash, but it returns some weirdass pointer that crashes applications.

To be precise, `resp->h_aliases[0]` contains 0x36312e30362e3031. Remember, it's a `char *`, and this value certainly doesn't seem to point anywhere. In fact, it is an ASCII string "61.06.01" somehow stuck into pointer value.

**Comment by WGH-** [comment]
```
#include <stdio.h>
#include <netdb.h>

int main(int argc, char **argv)
{
    struct in6_addr addr;
    struct hostent *resp;

    if (argc != 2) {
        fprintf(stderr, "Usage: %s addr\n", argv[0]);
        return 1;
    }

    if (!inet_pton(AF_INET6, argv[1], &addr)) {
        fprintf(stderr, "Bad address %s\n", argv[1]);
        return 1;
    }

    resp = gethostbyaddr(&addr, sizeof(addr), AF_INET6);

    if (resp) {
        printf("resp->h_name=%s\n", resp->h_name);
        if (resp->h_aliases) {
            char **pch;
            for (pch = resp->h_aliases; *pch != NULL; pch++) {
                printf("resp->h_aliases[i]=%p\n", *pch);
            }
        }
    } else {
        printf("res is NULL\n");
    }

    return 0;
}
```

**Comment by keszybz** [comment]
Oh, this should be fixed by 963783d7447cad48e594846a865b26535874679d.
